openlibrary/plugins/upstream/utils.py

```diff
@@ -709,51 +709,63 @@ def strip_accents(s: str) -> str:
     return {lang.key: lang for lang in web.ctx.site.get_many(keys)}
 
 
-def autocomplete_languages(prefix: str) -> Iterator[Storage]:
+def _normalize(s: str) -> str:
+    return strip_accents(s).lower()
+
+
+def _matches_lang_name(
+    prefix: str, language, translation: str | None = None
+) -> Storage | None:
+    lang_name = language.name
+    if translation is not None:
+        lang_name = safeget(lambda: language['name_translated'][translation][0])
+    # Compare to each inner word of language name for more accurate matching
+    # Eg. for prefix 'greek' will match with 'Ancient Greek' as well as 'Greek'
+    for inner in lang_name.split(' '):
+        if inner and _normalize(inner).startswith(prefix):
+            return Storage(
+                key=language.key,
+                code=language.code,
+                name=lang_name,
+            )
+    return None
+
+
+def autocomplete_languages(prefix: str, limit: int) -> Iterator[Storage]:
     """
-    Given, e.g., "English", this returns an iterator of:
+    Given, e.g., "English", this returns an iterator of the following up to given limit:
         <Storage {'key': '/languages/ang', 'code': 'ang', 'name': 'English, Old (ca. 450-1100)'}>
+        <Storage {'key': '/languages/cpe', 'code': 'cpe', 'name': 'Creoles and Pidgins, English-based (Other)'}>
         <Storage {'key': '/languages/eng', 'code': 'eng', 'name': 'English'}>
         <Storage {'key': '/languages/enm', 'code': 'enm', 'name': 'English, Middle (1100-1500)'}>
     """
 
-    def normalize(s: str) -> str:
-        return strip_accents(s).lower()
-
-    prefix = normalize(prefix)
+    matches = 0
+    prefix = _normalize(prefix)
     user_lang = web.ctx.lang or 'en'
-    for lang in get_languages().values():
-        user_lang_name = safeget(lambda: lang['name_translated'][user_lang][0])
-        if user_lang_name and normalize(user_lang_name).startswith(prefix):
-            yield Storage(
-                key=lang.key,
-                code=lang.code,
-                name=user_lang_name,
-            )
-            continue
-
-        lang_iso_code = safeget(lambda: lang['identifiers']['iso_639_1'][0])
-        native_lang_name = safeget(lambda: lang['name_translated'][lang_iso_code][0])
-        if native_lang_name and normalize(native_lang_name).startswith(prefix):
-            yield Storage(
-                key=lang.key,
-                code=lang.code,
-                name=native_lang_name,
-            )
-            continue
-
-        if normalize(lang.name).startswith(prefix):
-            yield Storage(
-                key=lang.key,
-                code=lang.code,
-                name=lang.name,
-            )
-            continue
+    for language in get_languages().values():
+        # ISO code used to translate the current languages name into its native name
+        lang_iso_code = safeget(lambda: language['identifiers']['iso_639_1'][0])
```

Comment:
I only wanted to add to the match logic rather than prune existing search functionality, but if this is determined to be unnecessary I'm all for cleaning it up further.

Reply:
That makes sense. It just caught my eye and seemed odd.